rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,385 @@
+
+For example, it seems wrong that a user can do this: `self.set_id(Index::default())`. And while this might not be that big of an issue ( "users should just *not* do that" ), what if we want to add other functionality onto widgets in the future? Maybe we add the ability to give widgets their own classes for theming. This might be confusing for the user who thinks that `self.add_class("foo")` should work on its own (but might not since the context also needs to  be made aware of the change).
+
+Additionally, `KayakContext`, has a lot of functionality that doesn't need to be exposed within the context of a widget, but needs to elsewhere (such as the `KayakContext::render` and `KayakContext::process_events` methods). Methods like these need to be public, but they're dangerous to use within a widget's render method.
+
+So in summary the two main issues are:
+
+1. Internal widget data is exposed to the user in the render function in a confusing manner
+2. Error-prone `KayakContext` methods are also exposed to the user in the render function when they shouldn't be
+
+## Solution
+
+I think the widget system could be restructured in a way that creates a layer of abstraction between the widgets and their context.
+
+Let's assume we have this widget:
+
+```rust
+#[widget]
+fn MyWidget(name: String, fill: Option<Color>) {
+  // ...
+}
+```
+
+### Props Extraction
+
+To get around the first issue of the widget having too much mutable access to its required widget data, we could separate the render function from the widget itself by placing the actual logic inside an associated method. 
+
+This would remove the access to a mutable `self`. But we still need access to the actual props. To solve this issue we can generate a new struct, `MyWidgetProps`, to contain the prop data. We can then pass this into the render function.
+
+And now we have something that looks a bit more like this:
+
+```rust
+struct MyWidget {
+  pub id: Index,
+  pub children: Children,
+  pub props: MyWidgetProps,
+  // ...
+}
+
+struct MyWidgetProps {
+  pub name: String,
+  pub fill: Option<Color>,
+}
+
+impl Widget for MyWidget {
+  fn render(&mut self, context: &mut KayakContext) {
+    Self::render(self.props, context);
+  }
+}
+
+impl MyWidget {
+  fn render(props: MyWidgetProps, context: &mut KayakContext) {
+    // Generated logic
+  }
+}
+```
+
+### Widget Context
+
+Now comes the second issue of direct access to `KayakContext`. I think a good way of solving this would be to generate an interface struct. This struct would act as a middleman between the render function and `KayakContext`, forwarding only select methods to the context.
+
+There are two ways of going about this.
+
+1. Create a local struct:
+
+```rust
+struct MyWidgetContext<'a> {
```

Comment:
I'm worried anytime I see: `<'a>` because it can mean that the lifetime trickles up and we end up with lifetime requirements that are annoying for end users to deal with.

Reply:
Does this context get created on render and then dropped after?

Reply:
> I'm worried anytime I see: `<'a>` because it can mean that the lifetime trickles up and we end up with lifetime requirements that are annoying for end users to deal with.

Yeah, but I'm not sure how else to split up the internal widget code from the user-facing API.

> Does this context get created on render and then dropped after?

Yeah, this is a temporary type that basically serves as an ephemeral interface to the widget and context. And can be used to track changes.

Reply:
I think the lifetime is fine here! 👍 Since it's only created and dropped we are not storing it anywhere we should be fine.